### ISSUE TITLE
Supporting other data types (e.g. video)

### ISSUE THREAD

**Issue by shakibyzn** [Helsinki-NLP/mammoth#53]
Hi,

Is it possible to use Mammoth for other seq2seq problems, such as multilingual video/image captioning? What I have in mind is to prepare video features in this format (batch, n_frames, emb_size) and read them using src_embeddings in the training script while keeping the rest unchanged in the config.yml (e.g., tgt_vocab).

**Comment by TimotheeMickus** [comment]
Hi!

Other input types are currently not implemented, although we've had this request more than once. We don't really have enough hands to look into it properly; hence up until now, we've focused on text-only applications.

It should however be feasible with a reasonably small amount of changes to the codebase, depending what you're exactly looking for. If a hacky solution and vanilla transformer layers are good enough, then you can try the following to train a model:

1. override the function for file reading here:
https://github.com/Helsinki-NLP/mammoth/blob/c6a193b1cc16bf7140520c44712bcf82701ec87d/mammoth/inputters/dataset.py#L33
and in particular adapt the closure `_make_dict` to retrieve your data properly:
https://github.com/Helsinki-NLP/mammoth/blob/c6a193b1cc16bf7140520c44712bcf82701ec87d/mammoth/inputters/dataset.py#L43-L51
How to do that concretely depends on how your data is formatted. 

2. tweak the data collator function here:
https://github.com/Helsinki-NLP/mammoth/blob/c6a193b1cc16bf7140520c44712bcf82701ec87d/mammoth/inputters/dataset.py#L157
Batched tensors are expected to be sequence first (so your features would need to end up in the format `(n_frames, batch_size, model_dim)`.)

3. turn off mapping input tokens to embeddings; cf. here for the encoder if I'm not mistaken:
https://github.com/Helsinki-NLP/mammoth/blob/c6a193b1cc16bf7140520c44712bcf82701ec87d/mammoth/modules/layer_stack_encoder.py#L92-L93

4.  use a default sentence-level batching function (`--batch_type sents`). You might need to also pass some dummy variables for the source vocab or comment out the relevant section.

If you'd like to have a look at doing that more properly, external contributions are very much welcome!

**Comment by shakibyzn** [comment]
Thank you for your detailed reply. I've been working on this for a month now and I'm able to train it properly. I haven't had a lot of experiments with that and I'm trying different hyperparameters to see if I can get to an acceptable performance. One question, isn't it possible to use loss as the criterion for early stopping?

**Comment by TimotheeMickus** [comment]
Yes, early stopping should be supported out of the box.

Assuming you want to evaluate every 10k steps, and stop training if it no longer improves after 5 evaluation loops, then:
1. provide a path_valid_src and a path_valid_tgt in your task definitions for enabling validation loops
2. include the following to your YAML config:
```yaml
early_stopping: 5
early_stopping_criteria: ppl
valid_steps: 10000
```

The early stopper will evaluate perplexity on the validation dataset(s), which should be equivalent to the cross-entropy used to train the model.
If you  need something finer-grained than that, then you can implement a Scorer object:

https://github.com/Helsinki-NLP/mammoth/blob/1ba7af94501647ae691db213f572360f89466f5d/mammoth/utils/earlystopping.py#L11

and then make sure to make it available as one of the default scorers and scorer builders:

https://github.com/Helsinki-NLP/mammoth/blob/1ba7af94501647ae691db213f572360f89466f5d/mammoth/utils/earlystopping.py#L60-L63

**Comment by shakibyzn** [comment]
Thank you.

**Comment by TimotheeMickus** [comment]
No worries! don't hesitate to share your code if you want us to include it in the library, we would welcome a pull request if you have something working.
